expr/func_timeshift.go

```diff
@@ -0,0 +1,153 @@
+	if err != nil {
+		// panic??? TODO - shouldn't happen, validated above
+		return context
+	}
+
+	// TODO - worry about underflow? uint32 can't represent date older than epoch
+	context.from = addOffset(context.from, s.shiftOffset)
+	context.to = addOffset(context.to, s.shiftOffset)
+
+	return context
+}
+
+func (s *FuncTimeShift) Exec(dataMap DataMap) ([]models.Series, error) {
+	series, err := s.in.Exec(dataMap)
+	if err != nil {
+		return nil, err
+	}
+
+	negativeOffset := -s.shiftOffset
+
+	formatStr := fmt.Sprintf("timeShift(%%s, \"%s\")", s.timeShift)
+	newName := func(oldName string) string {
+		return fmt.Sprintf(formatStr, oldName)
+	}
+
+	outputs := make([]models.Series, 0, len(series))
+	for _, serie := range series {
+		out := GetPooledSliceAtLeastSize(len(serie.Datapoints))
+		for _, v := range serie.Datapoints {
+			out = append(out, schema.Point{Val: v.Val, Ts: addOffset(v.Ts, negativeOffset)})
+		}
+
+		output := models.Series{
+			Target:       newName(serie.Target),
+			QueryPatt:    newName(serie.QueryPatt),
+			QueryFrom:    serie.QueryFrom,
+			QueryTo:      serie.QueryTo,
+			QueryMDP:     serie.QueryMDP,
+			QueryPNGroup: serie.QueryPNGroup,
+			Tags:         serie.CopyTagsWith("timeShift", s.timeShift),
+			Datapoints:   out,
+			Interval:     serie.Interval,
+			Meta:         serie.Meta,
+		}
+
+		outputs = append(outputs, output)
+	}
+	dataMap.Add(Req{}, outputs...)
+	return outputs, nil
+}
+
+func (s *FuncTimeShift) getTimeShiftSeconds(context Context) (int, error) {
+	// Trim off the sign (if there is one)
+	sign := -1
+	durStr := s.timeShift
+	if durStr[0] == '-' {
+		durStr = durStr[1:]
+	} else if durStr[0] == '+' {
+		sign = 1
+		durStr = durStr[1:]
+	} else {
+		// Normalize timeShift to always include the sign
+		s.timeShift = "-" + s.timeShift
+	}
+
+	interval, err := dur.ParseDuration(durStr)
+
+	if err != nil {
+		return 0, err
+	}
+
+	signedOffset := int(interval) * sign
+
+	if s.alignDST {
+		timezoneOffset := func(epochTime uint32) int {
+			// Use server timezone to determine DST
+			localTime := time.Now()
+			thenTime := time.Unix(int64(epochTime), 0)
+
+			_, offset := time.Date(thenTime.Year(), thenTime.Month(),
+				thenTime.Day(), thenTime.Hour(), thenTime.Minute(),
+				thenTime.Second(), thenTime.Nanosecond(), localTime.Location()).Zone()
+
+			return offset
+		}
+
+		reqStartOffset := timezoneOffset(context.from)
+		reqEndOffset := timezoneOffset(context.to)
+		myStartOffset := timezoneOffset(addOffset(context.from, signedOffset))
+		myEndOffset := timezoneOffset(addOffset(context.to, signedOffset))
+
+		dstOffset := 0
+
+		// Both requests ranges are entirely in the one offset (either in or out of DST)
+		if reqStartOffset == reqEndOffset && myStartOffset == myEndOffset {
+			dstOffset = reqStartOffset - myStartOffset
+		}
+
+		signedOffset += dstOffset
+	}
+
+	return signedOffset, nil
+}
+
+func addOffset(orig uint32, offset int) uint32 {
+	if offset < 0 {
+		orig -= uint32(offset * -1)
```

Comment:
so what happens if this result is negative (orig will be an incorrect positive value here)?

Reply:
Yes see comment at https://github.com/bloomberg/metrictank/pull/89/files/01ea8dd74d455203529e2456290e579b6cb12dc0#diff-a08ba41f2effbe7b76d8c3a9f891cd75R46